codec-redis/src/main/java/io/netty/handler/codec/redis/RedisDecoder.java

```diff
@@ -0,0 +1,238 @@
+        } else {
+            throw new IllegalArgumentException("bad length: " + current.length);
+        }
+    }
+
+    private boolean decodeBulkStrings(DecodeState current, ByteBuf in) throws Exception {
+        if (current.length == NULL_RESPONSE) {
+            // $-1\r\n
+            current.content = null;
+            current.state = State.DECODE_COMPLETE;
+        } else if (current.length == 0) {
+            // $0\r\n <here> \r\n
+            if (in.readableBytes() < CRLF_LENGTH) {
+                return false;
+            }
+            in.skipBytes(CRLF_LENGTH);
+            current.content = Unpooled.EMPTY_BUFFER;
+            current.state = State.DECODE_COMPLETE;
+        } else if (current.length > 0) {
+            // ${length}\r\n <here> {data...}\r\n
+            if (in.readableBytes() < current.length + CRLF_LENGTH) {
+                return false;
+            }
+            current.content = in.readSlice(current.length);
+            in.skipBytes(CRLF_LENGTH);
+            current.state = State.DECODE_COMPLETE;
+        } else {
+            throw new IllegalArgumentException("bad length: " + current.length);
+        }
+        return true;
+    }
+
+    private void decodeArrays(DecodeState current) throws Exception {
+        if (current.children.size() == current.length) {
+            current.state = State.DECODE_COMPLETE;
+        } else if (current.children.size() < current.length) {
+            stages.push(new DecodeState()); // for child
+        } else {
+            throw new IllegalArgumentException("children.size: " + current.children.size()
+                                               + ", current.length: " + current.length);
+        }
+    }
+
+    private void aggregate(DecodeState current, List<Object> out) throws Exception {
+        RedisMessage msg = newRedisMessage(current);
+        if (stages.size() == 1) {
+            out.add(msg);
+            stages.pop();
+            stages.push(new DecodeState()); // for next message
+        } else if (stages.size() > 1) {
+            stages.pop();
+            DecodeState parent = stages.peek();
+            parent.children.add(msg);
+        } else {
+            throw new Error("bad size: " + stages.size());
+        }
+    }
+
+    private static ByteBuf readLine(ByteBuf in) {
+        final int lfIndex = in.forEachByte(ByteProcessor.FIND_LF);
+        if (lfIndex < 0) {
+            return null;
+        }
+        final int length = lfIndex - in.readerIndex() - 1; // `-1` is for CR
+        return readBytes(in, length);
+    }
+
+    private static ByteBuf readBytes(ByteBuf in, int length) {
+        ByteBuf buf = in.readSlice(length);
+        in.skipBytes(CRLF_LENGTH); // skip CRLF
+        return buf;
+    }
+
+    private static int readLength(ByteBuf in) {
+        byte[] data = new byte[in.readableBytes()];
+        in.readBytes(data);
+        return Integer.parseInt(new String(data, CharsetUtil.UTF_8));
```

Comment:
This needs to be `Long` as Redis integers are base-10 64 bit signed integers which map to Java `long`